Json sig

lgautier opened the pull request:
Friends don't let friends use YAML as a data exchange format.

Reimplementing the current default YAML format to a JSON equivalent makes the loading of signatures 10 to 20 times faster while conserving a constant (and small) memory footprint by reading collections of signatures from a stream.

The benchmark is below.

```python
import sourmash_lib.signature

yaml_sig_filename = "microbe_sigs-2016-11-27-concat-head.sig"

with open(yaml_sig_filename) as fh:
    signatures = tuple(sourmash_lib.signature.load_signatures(fh))
     
with open(yaml_sig_filename + ".json", 'w') as fh:
    sourmash_lib.signature.save_signatures_json(signatures, fp=fh, indent=None, sort_keys=False)

import timeit

stmt = """
with open(filename) as fh:
    signatures = tuple(sourmash_lib.signature.load_signatures(fh))
"""
setup = "import sourmash_lib.signature; filename='%s'" % yaml_sig_filename)
t = timeit.Timer(stmt,
                 setup=setup)

print('YAML:')
bench_yaml = t.timeit(number=5)
print(bench_yaml)

print('---')

stmt = """
with open(filename, 'rb') as fh:
    signatures = tuple(sourmash_lib.signature_json.load_signatures_json(fh))
"""
setup = "import sourmash_lib.signature_json; filename='%s'" % (yaml_sig_filename + ".json"))
t = timeit.Timer(stmt,
                 setup=setup)

print('JSON:')
bench_json = t.timeit(number=5)
print(bench_json)

stmt = """
with open(filename, 'rb') as fh:
    signatures = tuple(sourmash_lib.signature_json.load_signatures_json(fh, ijson=ijson))
"""
setup = "import sourmash_lib.signature_json; import ijson.backends.yajl2_cffi as ijson; filename='%s'" % (yaml_sig_filename + ".json"))
t = timeit.Timer(stmt,
                 setup=setup)

print('JSON (faster backend):')
bench_json = t.timeit(number=5)
print(bench_json)
```

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests? (but the package `ijson` is required)
- [x] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

ctb commented:
Sorry, didn't notice the branch had been switched over - see #72 now instead ;)

lgautier commented:
Ah. Doesn't pass tests for Python 2.

Just let Python 2 users stuck with YAML as an incentive to move to Python 3.

lgautier commented:
The failed tests with Python 2 are around `str`/`unicode`. Seems trivial to fix **in the tests** but from experience this is a can of worms (and not worth time - Python is legacy and will reach end-of-support soon).

Code coverage is the tests appears good (90+ %) and the pull request can be merged as is with Python 3 in mind as it can be the basis for integration with new arguments for the command line. What to do with Python 2 (attempt to work out `str`/`unicode` glitches, or just make JSON unsupported - acceptable since YAML can be used to communicate between Python 2 and Python 3) is left to the reader (well, merger).

ctb commented:
#72 makes JSON the default output format.

#74 supports loading of YAML for legacy purposes, and also adds Python 2 compatibility.

I propose @lgautier review & merge those, and then we ask @luizirber to review this PR.

lgautier commented:
Git issue.

Looks like I nuked #74 merged into #71. I did not realize it until too late (I had accidentally merged the branch json-sig into my fork's master, and thought that I was the only person who cared about that fork).

Sorry about that.

ctb commented:
np at all!

lgautier commented:
Thanks @ctb . I am understanding that @luizirber will review before merge into the main repos' master (after which my mishaps with Git won't be able to cause any harm).

luizirber commented:
I think this PR is covered by #78, right?

ctb commented:
You can merge #78 and that should fix the CI.

Yes, I'd appreciate @luizirber thoughts, as he is the other knowledgeable person who has been thinking about formats.  Still, it's kind of a holiday week here in the US so if he doesn't get to it within a few days I'll just do a final review and merge it.  @betatim might poke his head in also.

ctb commented:
"covered by"?

On Mon, Jan 02, 2017 at 11:24:55AM -0800, Luiz Irber wrote:
> I think this PR is covered by #78, right?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/dib-lab/sourmash/pull/71#issuecomment-270011120
-- 
C. Titus Brown, ctbrown@ucdavis.edu

luizirber commented:
All the changes in this PR are in #78 too

ctb commented:
yes, I think there is some git confusion going on - I used the wrong base for
#78, but it's all in this one in any case.  If you can take a look at #71 I
think that's where the latest and greatest should go, so reviews here are
best.

lgautier commented:
> Yes, I'd appreciate @luizirber thoughts, as he is the other knowledgeable person who has been thinking about formats. Still, it's kind of a holiday week here in the US so if he doesn't get to it within a few days I'll just do a final review and merge it. @betatim might poke his head in also.

No hurry on my end. Just making sure I am not not merging something I should (yes, two "not").

luizirber commented:
OK, I will review this one

luizirber commented on sourmash_lib/signature_json.py:
```diff
@@ -0,0 +1,350 @@
+    return sig
+
+# integration test more than a unit test...
+def test__json_next_signature():
```
We can move tests to `sourmash_lib/test_signature.py`, but I can also do that later.

lgautier replied:
Whatever makes it **easier**, I'd say. Disclaimer: I just went down a rabbit hole with Git issues, and barely made it back. ;-)

lgautier commented on setup.py:
```diff
@@ -65,7 +65,7 @@
     "install_requires": ["screed>=0.9", "PyYAML>=3.11"],
     "extras_require": {
-        'test' : ['pytest', 'pytest-cov', 'numpy', 'matplotlib', 'scipy'],
+        'test' : ['pytest', 'pytest-cov', 'numpy', 'matplotlib', 'scipy', 'ijson'],
```
You are right. This was originally implemented as a very optional JSON format, but it now going to be the possible default.

lgautier commented:
I am just seeing that one of commit fixing tests for Python 2 broke them for Python 3 (initial PR only bothered about Python 3).

luizirber commented:
Thanks @lgautier! I'm merging this PR, next step is to reopen #78 and support both signatures (for now), @ctb?

ctb commented:
I believe #78 is included in #71, so we should be all good.

luizirber commented:
I don't think so, I don't see https://github.com/dib-lab/sourmash/pull/78/files#diff-48808513619dee1d863793d275d19f20R162 for example

lgautier commented:
@ctb, @luizirber : all checks "should" be passing. I have addressed both @luizirber review comments and fixed the Python2/3 issue with tests (more glitches with Python 2 may emerge though - this was written for Python 3, and I only used it with Python 3)

ctb commented:
reopened, then.  at one point it seemed like it was all in there :)

lgautier commented:
It almost certainly was. It was like playing whack-a-mole with git after my initial mistake.
